### PR TITLE
have batcher submit appropriate cancellation transactions when mempool is blocked

### PR DESCRIPTION
**Pull request by roberto-bayardo**
Description

Extend batcher to submit appropriate cancellation transactions when the mempool is blocked by a transaction of incompatible type.

Also makes the txmgr use the geth-defined AlreadyReservedErr instead of redefining it, now that op-geth has caught up to the commit that exposes it.

Tests

Extended the E2E tests to submit incompatible pending transactions to the txpool to confirm the new functionality appropriately clears them.

Metadata

- Fixes #[ethereum-optimism/protocol-quest/issues/102]

### CODE REVIEW

File: op-batcher/batcher/driver.go [hunk 1]
```diff
@@ -272,10 +294,21 @@ func (l *BatchSubmitter) loop() {
...
 		for {
 			select {
 			case r := <-receiptsCh:
+				if errors.Is(r.Err, txpool.ErrAlreadyReserved) && txpoolState.CompareAndSwap(TxpoolGood, TxpoolBlocked) {
+					l.Log.Info("incompatible tx in txpool")
+				} else if r.ID[0].chID == cancelID &&
+					txpoolState.CompareAndSwap(TxpoolCancelPending, TxpoolGood) {
+					// Set state to TxpoolGood even if the cancellation transaction ended in error
+					// since the stuck transaction could have cleared while we were waiting.
+					l.Log.Info("txpool may no longer be blocked", "err", r.Err)
```
Comment (axelKingsley) [inline op-batcher/batcher/driver.go]: Just to check my understanding -- this indicates that the manager may send the cancellation tx many times if it ends in error, correct? the pool state would rotate to "Good", then if it is still not cleared, would cycle back through to blocked?

File: op-batcher/batcher/driver.go [hunk 2]
```diff
@@ -272,10 +294,21 @@ func (l *BatchSubmitter) loop() {
...
 	defer close(receiptLoopDone) // shut down receipt loop
+
+	var txpoolState atomic.Int32
+	txpoolState.Store(TxpoolGood)
 	go func() {
 		for {
 			select {
 			case r := <-receiptsCh:
+				if errors.Is(r.Err, txpool.ErrAlreadyReserved) && txpoolState.CompareAndSwap(TxpoolGood, TxpoolBlocked) {
+					l.Log.Info("incompatible tx in txpool")
```
Comment (axelKingsley) [inline op-batcher/batcher/driver.go:313]: Having the swap be part of the condition is technically fine, but makes it a little hard to follow -- short-circuiting from the `errors.Is` is how this check avoids setting the state to "Blocked", but could we put that *inside* the condition? Are there cases where the error is `ErrAlreadyReserved`, but the state is *not* "Good"?

File: op-batcher/batcher/driver.go [hunk 3]
```diff
@@ -479,6 +521,32 @@ func (l *BatchSubmitter) safeL1Origin(ctx context.Context) (eth.BlockID, error)
...
+// cancelBlockingTx creates an empty transaction of appropriate type to cancel out the incompatible
+// transaction stuck in the txpool. In the future we might send an actual batch transaction instead
+// of an empty one to avoid wasting the tx fee.
+func (l *BatchSubmitter) cancelBlockingTx(queue *txmgr.Queue[txID], receiptsCh chan txmgr.TxReceipt[txID]) {
+	txdata := txData{
+		frames: []frameData{
+			frameData{
+				data: []byte{},
+				id: frameID{
+					chID:        cancelID,
```
Comment (axelKingsley) [inline op-batcher/batcher/driver.go]: I don't know offhand, is there any danger to having ID collisions if multiple channels end up getting cancelled?

File: op-batcher/batcher/driver.go [hunk 4]
```diff
@@ -515,6 +583,11 @@ func (l *BatchSubmitter) sendTransaction(ctx context.Context, txdata txData, que
 		candidate = l.calldataTxCandidate(data)
 	}
 
+	l.queueTx(txdata, candidate, queue, receiptsCh)
+	return nil
+}
+
+func (l *BatchSubmitter) queueTx(txdata txData, candidate *txmgr.TxCandidate, queue *txmgr.Queue[txID], receiptsCh chan txmgr.TxReceipt[txID]) {
```
Comment (axelKingsley) [inline op-batcher/batcher/driver.go]: Nice extraction :)

File: op-batcher/batcher/driver.go [hunk 5]
```diff
@@ -606,3 +678,50 @@ func logFields(xs ...any) (fs []any) {
 	}
 	return fs
 }
+
+// JamTxPool is for testing ONLY. It sends a txpool-blocking transaction in a new (non-blocking)
+// goroutine to the txpool that will panic if the transaction completes successfully. (Tests should
+// expect the blocking transaction to result in error from being cancelled.) This function must be
+// called *before* the batcher starts submitting batches to ensure successful jamming, and will
+// panic otherwise.
+func (l *BatchSubmitter) JamTxPool(ctx context.Context) {
```
Comment (axelKingsley) [inline op-batcher/batcher/driver.go]: Could `JamTxPool` call an internally-held `jamTxPoolFunc` that is stored in the struct? That way, production BatchSubmitters would have no way of triggering panics, while test-BatchSubmitters could be given this testing behavior.

File: op-service/txmgr/send_state.go [hunk 6]
```diff
@@ -76,7 +72,7 @@ func (s *SendState) ProcessSendError(err error) {
 		s.successFullPublishCount++
 	case errStringMatch(err, core.ErrNonceTooLow):
 		s.nonceTooLowCount++
-	case errStringMatch(err, ErrAlreadyReserved):
+	case errStringMatch(err, txpool.ErrAlreadyReserved):
 		s.alreadyReserved = true
```
Comment (axelKingsley) [inline op-service/txmgr/send_state.go:76]: Other cases here seem to count up the errors by type, while this just counts true/false. Any reason for that?

File: op-batcher/batcher/driver.go (hunk 2 above)
Comment (axelKingsley) [inline op-batcher/batcher/driver.go:313]: Why not just send the cancellation at this point, rather than waiting for another cycle through the loop? It goes from Good->Blocked, and then from Blocked->Pending, when it could maybe just go from Good->Pending?

**Review by axelKingsley — COMMENTED** [review]
This solution manages an enumerated State of the Transaction Pool, specifically whether it is unblocked, blocked, or pending(cancel). It does this by detecting a specific error type which sets the state to Blocked. When blocked it sends and empty batch with a special ID, and progresses to Pending. When in Pending, it optimistically advances back to Unblocked.

I left some comments to better understand the behavior and suggest some potential simplifications. This solution seems reasonable given the scope of changes, though it'd be nice to have a more comprehensive handling as part of the architecture at some point down the road.

File: op-batcher/batcher/driver.go (hunk 2 above)
Comment (roberto-bayardo) [reply op-batcher/batcher/driver.go:313]: Not sure I follow the concern about readability.  This seemed a natural way to me to have it print out the warning exactly once per cancellation attempt. Perhaps related: yes, ErrAlreadyReserved can be returned in any of the three states.  Imagine having published a bunch of transactions all at once of type calldata when there is a blocking blob tx.  Each would likely result in ErrAlreadyReserved here, with only the first receipt flipping the state from Good to Blocked.  For the remaining txs, the state would most likely be Blocked (but could be pending or even good depending on the timing of the cancellation tx handling.)

Comment (roberto-bayardo) [reply op-batcher/batcher/driver.go:313]: I thought it would be best to keep this loop purely handling receipts, and the other goroutine purely for sending. Mixing these responsibilities makes me a little nervous though I suppose it could work?  There could be potential blocking going on in the tx publishing routines, e.g. if queues get full for example, that might block receipt processing?

File: op-batcher/batcher/driver.go (hunk 1 above)
Comment (roberto-bayardo) [reply op-batcher/batcher/driver.go]: Correct. the state might flip to Good here, but then a later publish attempt could again result in "ErrAlreadyReserved" which would kick off another cancellation attempt.

File: op-batcher/batcher/driver.go (hunk 3 above)
Comment (roberto-bayardo) [reply op-batcher/batcher/driver.go]: Channels aren't getting cancelled, a lingering tx of incompatible type is what we're trying to cancel (via replacement).  There can only be one cancellation tx in flight at any given moment, so there should be no collisions from using this special ID here. (If I'm understanding you correctly.)

File: op-batcher/batcher/driver.go (hunk 5 above)
Comment (roberto-bayardo) [reply op-batcher/batcher/driver.go]: Not sure I follow, though I'm all ears if there's a clever way to prevent this from being used accidentally in prod.

File: op-service/txmgr/send_state.go (hunk 6 above)
Comment (roberto-bayardo) [reply op-service/txmgr/send_state.go:76]: It's not an error we expect to recur beyond the initial (successful) cancellation, so this seemed more natural.  though you're right looking at successFullPublishCount it's a bit inconsistent. Personally I think that one should also be changed to a bool :).     nonceTooLow must be a counter since we actually rely on the value being more than just 0/non-zero.

File: op-batcher/batcher/driver.go (hunk 5 above)
Comment (axelKingsley) [reply op-batcher/batcher/driver.go]: I mean that the Batch Submitter struct could contain,

```go
type BatchSubmitter struct {
...
jamPoolFn    func(context.Context)
...
}
```
And then this function you have here would proxy through to the stored function,
```go
func (l *BatchSubmitter) JamTxPool(ctx context.Context){
    if l.jamPoolFn != nil {
        l.jamPoolFn(ctx)
    }
    // implied else: no function set, nothing to do
}
```

This way, even calling this test function won't have any production effects. You'd set the function on the Batch Submitter in the tests that use it.

This is very optional feedback, and IDK if others would even agree with this approach.

File: op-batcher/batcher/driver.go (hunk 2 above)
Comment (axelKingsley) [reply op-batcher/batcher/driver.go:313]: That's a fair point, the status chain helps to keep the threads isolated by purpose :+1:

File: op-batcher/batcher/driver.go (hunk 5 above)
Comment (roberto-bayardo) [reply op-batcher/batcher/driver.go]: Ah I see. The jamming function needs to access the internal / non-exported methods of BatchSubmitter though so the func couldn't be defined outside the package. So not sure that buys us that much?

**Comment by roberto-bayardo** [comment]
I have rebased the PR to clear the conflicts.

File: op-batcher/batcher/driver.go (hunk 5 above)
Comment (sebastianst) [reply op-batcher/batcher/driver.go]: Maybe just to hide this a bit from the prod `BatchSubmitter`, could we embed the `BatchSubmitter` into a `TestBatchSubmitter` struct and then expose such functions only on this `TestBatchSubmitter`? We could then add a `TestDriver() *TestBatchSubmitter` getter to the `BatcherService` and access it in `op-e2e/eip4844_test.go` where `action` is defined.

File: op-batcher/batcher/driver.go [hunk 7]
```diff
@@ -606,3 +679,50 @@ func logFields(xs ...any) (fs []any) {
...
```
Comment (sebastianst) [inline op-batcher/batcher/driver.go]: Instead of calling panics inside `JamTxPool` and spawning a go routine, could we instead make this function return `error` and synchronously call the actual sending code, and then in the test we could run this function in a goroutine and we could capture the returned error in a channel and assert on it. This would avoid panics, and give the test the ability to own the goroutine.

File: op-service/txmgr/txmgr.go [hunk 8]
```diff
@@ -905,3 +905,71 @@ func finishBlobTx(message *types.BlobTx, chainID, tip, fee, blobFee, value *big.
 	}
 	return nil
 }
+
+// JamTxPool sends a transaction intended to get stuck in the txpool, and should be used ONLY for testing.
+func (m *SimpleTxManager) JamTxPool(ctx context.Context, candidate TxCandidate) (*types.Receipt, error) {
```
Comment (sebastianst) [inline op-service/txmgr/txmgr.go]: Similarly, we could add a `TestTxManager` with such methods and then only in the `TestBatchSubmitter` give access to it, to hide `JamTxPool` from the prod `SimpleTxManager`.

File: op-batcher/batcher/driver.go [hunk 9]
```diff
@@ -18,11 +21,17 @@ import (
...
 	"github.com/ethereum/go-ethereum/core/types"
 	"github.com/ethereum/go-ethereum/log"
 )
 
-var ErrBatcherNotRunning = errors.New("batcher is not running")
+var (
+	ErrBatcherNotRunning = errors.New("batcher is not running")
+
+	// cancelID is the channel ID used to indicate this transaction is a cancellation transaction only
+	cancelID = derive.ChannelID([derive.ChannelIDLength]byte{0xDE, 0xAD, 0xBE, 0xEF, 0xDE, 0xAD, 0xBE, 0xEF, 0xDE, 0xAD, 0xBE, 0xEF, 0xDE, 0xAD, 0xBE, 0xEF})
```
Comment (sebastianst) [inline op-batcher/batcher/driver.go]: Instead of having a magic cancel channel ID, we could define an augmented type around `txID` like
```go
type txRef struct {
    id txID
    isCancel bool
}
```
and then in the driver change the receipts channel and queue to
```go
	receiptsCh := make(chan txmgr.TxReceipt[txRef])
	queue := txmgr.NewQueue[txRef](l.killCtx, l.Txmgr, l.Config.MaxPendingTransactions)
```
so we can more clearly communicate our intent with this bool field instead of a magic channel id.

File: op-batcher/batcher/driver.go [hunk 10]
```diff
@@ -478,6 +521,32 @@ func (l *BatchSubmitter) safeL1Origin(ctx context.Context) (eth.BlockID, error)
...
+		frames: []frameData{
+			frameData{
+				data: []byte{},
+				id: frameID{
+					chID:        cancelID,
+					frameNumber: 0,
+				},
+			},
+		},
+	}
```
Comment (sebastianst) [inline op-batcher/batcher/driver.go]: ... we wouldn't then need to create any dummy data here...

File: op-batcher/batcher/driver.go [hunk 11]
```diff
@@ -271,10 +294,21 @@ func (l *BatchSubmitter) loop() {
...
+
+	var txpoolState atomic.Int32
+	txpoolState.Store(TxpoolGood)
 	go func() {
 		for {
 			select {
 			case r := <-receiptsCh:
+				if errors.Is(r.Err, txpool.ErrAlreadyReserved) && txpoolState.CompareAndSwap(TxpoolGood, TxpoolBlocked) {
+					l.Log.Info("incompatible tx in txpool")
+				} else if r.ID[0].chID == cancelID &&
```
Comment (sebastianst) [inline op-batcher/batcher/driver.go]: ...and could here just test
```suggestion
				} else if r.isCancel &&
```

File: op-batcher/batcher/driver.go (hunk 5 above)
Comment (roberto-bayardo) [reply op-batcher/batcher/driver.go]: @sebastianst nice idea, done!

File: op-service/txmgr/txmgr.go (hunk 8 above)
Comment (roberto-bayardo) [reply op-service/txmgr/txmgr.go]: done

File: op-batcher/batcher/driver.go (hunk 7 above)
Comment (roberto-bayardo) [reply op-batcher/batcher/driver.go]: done

Comment (roberto-bayardo) [reply op-batcher/batcher/driver.go]: done, though this identified a race condition where the batcher can start before the jamming goroutine starts executing.  Fixed for now by inserting a 1 second sleep to give it time to start up. Another solution could be to just use a short timeout so that the jamming function doesn't block for too long, allowing it to be invoked without its own goroutine without blocking the entire test, but I think that's just effectively doing the same thing as an explicit sleep.

Comment (roberto-bayardo) [reply op-batcher/batcher/driver.go]: I can fix this by separating out the send from the waiting in txmgr... let me do that.

Comment (roberto-bayardo) [reply op-batcher/batcher/driver.go]: this is now done

File: op-batcher/batcher/driver.go (hunk 9 above)
Comment (roberto-bayardo) [reply op-batcher/batcher/driver.go]: Was hoping avoid adding another indirection here, and saw that channelID was 128 bits and therefore basically collision proof so opted for using a cahnnelID indicator for cancel.  Don't feel so strongly so happy to change it if you prefer it.

File: op-batcher/batcher/driver.go (hunk 10 above)
Comment (roberto-bayardo) [reply op-batcher/batcher/driver.go]: looks like I still need some dummy data to create the tx candidate

File: op-batcher/batcher/driver.go (hunk 11 above)
Comment (roberto-bayardo) [reply op-batcher/batcher/driver.go]: done

File: op-batcher/batcher/driver.go (hunk 9 above)
Comment (roberto-bayardo) [reply op-batcher/batcher/driver.go]: this is now done.

File: op-batcher/batcher/driver.go [hunk 12]
```diff
@@ -606,3 +676,60 @@ func logFields(xs ...any) (fs []any) {
 	}
 	return fs
 }
+
+type TestBatchSubmitter struct {
+	bs  *BatchSubmitter
```
Comment (sebastianst) [inline op-batcher/batcher/driver.go]:
```suggestion
	*BatchSubmitter
```
If we embed, all methods are automatically available and we don't need to forward e.g. `StartBatchSubmitting`. Or did you deliberately not embed?

File: op-batcher/batcher/service.go [hunk 13]
```diff
@@ -428,6 +428,14 @@ var _ cliapp.Lifecycle = (*BatcherService)(nil)
...
 // Driver returns the handler on the batch-submitter driver element,
 // to start/stop/restart the batch-submission work, for use in testing.
-func (bs *BatcherService) Driver() rpc.BatcherDriver {
+func (bs *BatcherService) Driver() *BatchSubmitter {
 	return bs.driver
 }
+
+// NewTestDriver returns a handler for the batch-submitter driver element, to start/stop/restart
+// the batch-submission work, for use only in testing.
+func (bs *BatcherService) NewTestDriver() *TestBatchSubmitter {
```
Comment (sebastianst) [inline op-batcher/batcher/service.go]: nit, to be consistent with `Driver`
```suggestion
func (bs *BatcherService) TestDriver() *TestBatchSubmitter {
```

File: op-e2e/eip4844_test.go [hunk 14]
```diff
@@ -50,13 +51,37 @@ func testSystem4844E2E(t *testing.T, multiBlob bool) {
...
+	// transaction. The hook used here makes sure we make the jamming call before batch submission
+	// is started, as is required by the function.
+	jamChan := make(chan error)
+	jamCtx, jamCancel := context.WithTimeout(context.Background(), 20*time.Second)
+	action := SystemConfigOption{
+		key: "beforeBatcherStart",
+		action: func(cfg *SystemConfig, s *System) {
+			driver := s.BatchSubmitter.NewTestDriver()
+			err := driver.JamTxPool(jamCtx)
+			require.Nil(t, err)
```
Comment (sebastianst) [inline op-e2e/eip4844_test.go]:
```suggestion
			require.NoError(t, err)
```
better failure messages...

File: op-e2e/eip4844_test.go [hunk 15]
```diff
@@ -50,13 +51,37 @@ func testSystem4844E2E(t *testing.T, multiBlob bool) {
...
+			go func() {
+				jamChan <- driver.WaitOnJammingTx(jamCtx)
+			}()
+		},
+	}
+	defer func() {
+		jamCancel()
+		require.Nil(t, <-jamChan)
```
Comment (sebastianst) [inline op-e2e/eip4844_test.go]:
```suggestion
		require.NoError(t, <-jamChan, "jam tx error")
```
probably nice to have some context in case this error is non-nil

File: op-service/txmgr/txmgr.go [hunk 16]
```diff
@@ -905,3 +905,87 @@ func finishBlobTx(message *types.BlobTx, chainID, tip, fee, blobFee, value *big.
 	}
 	return nil
 }
+
+type TestTxManager struct {
```
Comment (sebastianst) [inline op-service/txmgr/txmgr.go]: would be nice to move this type and its methods into a new file `test_txmgr.go`.

**Review by sebastianst — COMMENTED** [review]
Looks good! Some minor improvement suggestions.

File: op-batcher/batcher/driver.go (hunk 12 above)
Comment (roberto-bayardo) [reply op-batcher/batcher/driver.go]: I had it embedded originally but it doesn't work because then you end up with copies of the variables whose changes aren't seen by the "main" driver, and it breaks several tests. For example calling "StopBatcher" on the TestDriver wasn't actually stopping the batcher.

Comment (sebastianst) [reply op-batcher/batcher/driver.go]: I see, but what about embedding a pointer?

Comment (roberto-bayardo) [reply op-batcher/batcher/driver.go]: good point, done!

File: op-service/txmgr/txmgr.go (hunk 16 above)
Comment (roberto-bayardo) [reply op-service/txmgr/txmgr.go]: done.. should we also move TestBatchSubmitter into its own file?

File: op-e2e/eip4844_test.go (hunk 15 above)
Comment (roberto-bayardo) [reply op-e2e/eip4844_test.go]: done

File: op-e2e/eip4844_test.go (hunk 14 above)
Comment (roberto-bayardo) [reply op-e2e/eip4844_test.go]: done

File: op-service/txmgr/txmgr.go (hunk 16 above)
Comment (sebastianst) [reply op-service/txmgr/txmgr.go]: good idea! sorry, missed that

Comment (roberto-bayardo) [reply op-service/txmgr/txmgr.go]: moved TestBatchSubmitter into its own file too

**Review by sebastianst — APPROVED** [review]
